File: apps/android/app/src/main/java/com/meshx/android/MeshXAuthPlugin.java
package com.meshx.android;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.RejectedExecutionException;

/** Capacitor boundary for Android native cookie-based authentication. */
@CapacitorPlugin(name = "MeshXAuth")
public class MeshXAuthPlugin extends Plugin {
    private MeshXAuthClient client;
    private ExecutorService executor;

    @Override
    public void load() {
        client = new MeshXAuthClient(new EncryptedOriginSessionStore(getContext()));
        // Blocking OkHttp calls (up to the 12s call timeout) must stay off
        // Capacitor's shared plugin thread. A single-thread executor also
        // serializes requests, so the client's in-flight guard keeps working.
        executor = Executors.newSingleThreadExecutor();
    }

    @Override
    protected void handleOnDestroy() {
        if (executor != null) executor.shutdownNow();
    }

    @PluginMethod
    public void login(PluginCall call) {
        executeExclusive(call, () -> {
            MeshXAuthClient.AuthSession session = client.login(
                    call.getString("origin"),
                    call.getString("apiBasePath", MeshXAuthClient.API_BASE_PATH),
                    call.getString("username"),
                    call.getString("password"),
                    call.getString("deviceName", "MeshX Android"));
            call.resolve(sessionObject(session));
        });
    }

    @PluginMethod
    public void refresh(PluginCall call) {
        executeExclusive(call, () -> {
            MeshXAuthClient.AuthSession session = client.refresh(
                    call.getString("origin"),
                    call.getString("apiBasePath", MeshXAuthClient.API_BASE_PATH),
                    call.getString("deviceName", "MeshX Android"));
            call.resolve(sessionObject(session));
        });
    }

    @PluginMethod
    public void logout(PluginCall call) {
        executeExclusive(call, () -> {
            client.logout(
                    call.getString("origin"),
                    call.getString("apiBasePath", MeshXAuthClient.API_BASE_PATH),
                    call.getString("accessToken"));
            call.resolve();
        });
    }

    @PluginMethod
    public void clearNodeSession(PluginCall call) {
        execute(call, () -> {
            client.clearNodeSession(call.getString("origin"));
            call.resolve();
        });
    }

    /**
     * The single-thread executor queues submissions, so the client's
     * requireIdle guard can never observe a busy state from a call routed
     * through it: a second request would silently wait behind an in-flight
     * one (up to the 12s call timeout) instead of failing fast. Capacitor
     * dispatches plugin methods from a single thread, so this check-then-submit
     * cannot race; requireIdle stays in the client as the defensive invariant.
     */
    private void executeExclusive(PluginCall call, AuthAction action) {
        if (client.isBusy(call.getString("origin"))) {
            call.reject("a native authentication request is already in progress", "AUTH_BUSY");
            return;
        }
        execute(call, action);
    }

    private void execute(PluginCall call, AuthAction action) {
        try {
            executor.execute(() -> {
                try {
                    action.run();
                } catch (Exception exception) {
                    reject(call, exception);
                }
            });
        } catch (RejectedExecutionException exception) {
            reject(call, exception);
        }
    }

    private static JSObject sessionObject(MeshXAuthClient.AuthSession session) {
        JSObject result = new JSObject();
        result.put("userId", session.userId());
        result.put("username", session.username());
        result.put("nickname", session.nickname());
        if (session.avatar() != null) result.put("avatar", session.avatar());
        result.put("token", session.token());
        result.put("expiresIn", session.expiresIn());
        return result;
    }

    private static void reject(PluginCall call, Exception exception) {
        String message = exception.getMessage();
        String text = message == null || message.isBlank()
                ? "Android native authentication failed"
                : message;
        if (exception instanceof MeshXAuthClient.BusyException) {
            call.reject(text, "AUTH_BUSY");
        } else {
            call.reject(text);
        }
    }

    @FunctionalInterface
    private interface AuthAction {
        void run() throws Exception;
    }
}

File: apps/ios/ios/App/App/MeshXAuthPlugin.swift
import Capacitor
import Foundation

/// JS 只拿 access token、用户资料与 expiresIn；refresh token 留在原生层。
/// 忙碌时快速失败（AUTH_BUSY），与 Android 插件语义一致。
@objc(MeshXAuthPlugin)
public class MeshXAuthPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MeshXAuthPlugin"
    public let jsName = "MeshXAuth"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "login", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "refresh", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "logout", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearNodeSession", returnType: CAPPluginReturnPromise),
    ]

    private let client = MeshXAuthClient()
    private let queue = DispatchQueue(label: "com.atti20.lanchat.meshx-auth", qos: .userInitiated)

    @objc func login(_ call: CAPPluginCall) {
        guard let origin = call.getString("origin"),
              let apiBasePath = call.getString("apiBasePath"),
              let username = call.getString("username"),
              let password = call.getString("password"),
              let deviceName = call.getString("deviceName") else {
            call.reject("login requires origin, apiBasePath, username, password and deviceName")
            return
        }
        run(call, origin: origin) {
            let session = try self.client.login(
                rawOrigin: origin,
                apiBasePath: apiBasePath,
                username: username,
                password: password,
                deviceName: deviceName
            )
            call.resolve(Self.sessionPayload(session))
        }
    }

    @objc func refresh(_ call: CAPPluginCall) {
        guard let origin = call.getString("origin"),
              let apiBasePath = call.getString("apiBasePath"),
              let deviceName = call.getString("deviceName") else {
            call.reject("refresh requires origin, apiBasePath and deviceName")
            return
        }
        run(call, origin: origin) {
            let session = try self.client.refresh(
                rawOrigin: origin,
                apiBasePath: apiBasePath,
                deviceName: deviceName
            )
            call.resolve(Self.sessionPayload(session))
        }
    }

    @objc func logout(_ call: CAPPluginCall) {
        guard let origin = call.getString("origin"),
              let apiBasePath = call.getString("apiBasePath") else {
            call.reject("logout requires origin and apiBasePath")
            return
        }
        let accessToken = call.getString("accessToken")
        run(call, origin: origin) {
            try self.client.logout(rawOrigin: origin, apiBasePath: apiBasePath, accessToken: accessToken)
            call.resolve()
        }
    }

    @objc func clearNodeSession(_ call: CAPPluginCall) {
        guard let origin = call.getString("origin") else {
            call.reject("clearNodeSession requires origin")
            return
        }
        queue.async {
            do {
                try self.client.clearNodeSession(rawOrigin: origin)
                call.resolve()
            } catch let error as MeshXAuthError {
                call.reject(error.message)
            } catch {
                call.reject("unable to clear the native node session")
            }
        }
    }

    private func run(_ call: CAPPluginCall, origin: String, _ operation: @escaping () throws -> Void) {
        if client.isBusy(rawOrigin: origin) {
            call.reject("a native authentication request is already in progress", "AUTH_BUSY")
            return
        }
        queue.async {
            do {
                try operation()
            } catch let error as MeshXAuthError {
                switch error {
                case .busy(let message):
                    call.reject(message, "AUTH_BUSY")
                case .failure(let message):
                    call.reject(message)
                }
            } catch {
                call.reject("native authentication request failed")
            }
        }
    }

    /// 载荷键集合固定为 {userId, username, nickname, avatar?, token, expiresIn}。
    private static func sessionPayload(_ session: MeshXAuthSession) -> [String: Any] {
        var payload: [String: Any] = [
            "userId": session.userId,
            "username": session.username,
            "nickname": session.nickname,
            "token": session.token,
            "expiresIn": session.expiresIn,
        ]
        if let avatar = session.avatar { payload["avatar"] = avatar }
        return payload
    }
}
